comp/core/workloadmeta/collectors/internal/remote/generic.go

```diff
@@ -113,9 +113,17 @@ func (c *GenericCollector) Start(ctx context.Context, store workloadmeta.Compone
 		opts = append(opts, grpc.WithTransportCredentials(creds))
 	}
 
+	opts = append(opts, grpc.WithContextDialer(func(_ context.Context, url string) (net.Conn, error) {
+		if strings.HasPrefix(url, "/") {
+			return net.Dial("unix", url)
+		}
+
+		return net.Dial("tcp", url)
+	}))
```

Comment:
I suggest making the generic implementation branch based on the URL provided via the address. I think we could probably create a new public function in the generic interface that each collector has to implement to return the correct connection struct

Reply:
I can definitely add a `NewConnection` or `Connect` function to the generic interface.
Just to make sure I understand, why not modifying the generic implementation ? The socket vs tcp is an implementation detail.

Reply:
> Just to make sure I understand, why not modify the generic implementation? The socket vs tcp is an implementation detail.

The way I see it, the generic wrapper simply calls the function that the different implementations have. So instead of checking the return value of `Address()` we can abstract that and use something like:

```go
opts := []grpc.DialOption{grpc.WithContextDialer(func(_ context.Context, url string) (net.Conn, error) {
    return c.StreamHandler.NewConnection(url)
})}
```

The way the logic of which type of connection is pushed to each specific implementation.

Of course this is a possible way of doing it, and the way you've done it is valid as well 😄 But If I were making the change, I would go for `NewConnection`

Reply:
We talked over Slack and agree that @lebauce point are correct and we do not to duplicate the logic in the NewConnection 😄